### PR TITLE
Add progress monitor

### PR DESCRIPTION
**Pull request by arjunmajum**
- Adds a progress sensor.
- Adds a global auxiliary loss registry.
- Adds a progress estimation network and loss term.

### CODE REVIEW

File: habitat/tasks/vln/vln.py [hunk 1]
```diff
@@ -131,6 +131,45 @@ def get_observation(
...
+    def get_observation(
+        self, observations, *args: Any, episode, **kwargs: Any
+    ):
+        current_position = self._sim.get_agent_state().position.tolist()
+
+        distance_to_target = self._sim.geodesic_distance(
+            current_position, episode.goals[0].position
+        )
+
+        distance_from_start = self._sim.geodesic_distance(
```
Comment (erikwijmans) [inline habitat/tasks/vln/vln.py]: This value is also on the episode under `episode.info["geodesic_distance"]`

File: habitat_baselines/models/rcm/vln_rcm_policy.py [hunk 2]
```diff
@@ -250,6 +271,16 @@ def forward(self, observations, rnn_hidden_states, prev_actions, masks):
             [state, text_embedding, rgb_embedding, depth_embedding], dim=1
         )
 
+        progress_hat = torch.tanh(self.progress_monitor(x))
```
Comment (erikwijmans) [inline habitat_baselines/models/rcm/vln_rcm_policy.py]: This can all go under the if guard -- that way you don't need the progress sensor enabled at all during eval.

File: habitat_baselines/models/rcm/vln_rcm_policy.py [hunk 3]
```diff
@@ -164,6 +165,16 @@ def __init__(
             "_scale", torch.tensor(1.0 / ((hidden_size // 2) ** 0.5))
         )
 
+        self.progress_monitor = nn.Linear(
+            self.vln_config.STATE_ENCODER.hidden_size
```
Comment (erikwijmans) [inline habitat_baselines/models/rcm/vln_rcm_policy.py]: This can all be `self.output_size`

File: habitat_baselines/models/vln_baseline_policy.py [hunk 4]
```diff
@@ -131,4 +151,15 @@ def forward(self, observations, rnn_hidden_states, prev_actions, masks):
             [instruction_embedding, depth_embedding, rgb_embedding], dim=1
         )
         x, rnn_hidden_states = self.state_encoder(x, rnn_hidden_states, masks)
+
+        progress_hat = torch.tanh(self.progress_monitor(x))
+        progress_loss = F.mse_loss(progress_hat.squeeze(1), observations['progress'])
+
+        if self.vln_config.PROGRESS_MONITOR.use and AuxLosses.is_active():
```
Comment (erikwijmans) [inline habitat_baselines/models/vln_baseline_policy.py:156]: Same comment about if statement.

**Review by erikwijmans — CHANGES_REQUESTED** [review]

**Comment by arjunmajum** [comment]
@erikwijmans made the changes, thanks for the review!

File: habitat_baselines/il/vln/dagger.py [hunk 5]
```diff
@@ -630,20 +643,32 @@ def train(self) -> None:
...
                             f"On DAgger iter {dagger_it}, Epoch {epoch}."
                         )
                         writer.add_scalar(
                             f"train_loss_iter_{dagger_it}", loss, step_id
                         )
+                        writer.add_scalar(
+                            f"train_action_loss_iter_{dagger_it}", action_loss, step_id
+                        )
+                        writer.add_scalar(
+                            f"train_aux_loss_iter_{dagger_it}", aux_loss, step_id
```
Comment (jacobkrantz) [inline habitat_baselines/il/vln/dagger.py:665]: In the event we had multiple aux losses, we would likely want to monitor them separately. Getting a dict from `AuxLosses` could do that for us. I don't think we plan on multiple aux losses, just a general thought.

**Review by jacobkrantz — COMMENTED** [review]
The [paper](https://arxiv.org/pdf/1901.03035.pdf) modifies inference when using PM:
> If the progress monitor output decreases (...), the agent is required to move back to the previous viewpoint and select the action with next highest probability.

Is there a reason we wouldn't want to include progress inference?

File: habitat_baselines/il/vln/dagger.py (hunk 5 above)
Comment (arjunmajum) [reply habitat_baselines/il/vln/dagger.py:665]: Yes, agreed. We should updated this when we have another auxiliary loss.

File: habitat_baselines/models/rcm/vln_rcm_policy.py [hunk 6]
```diff
@@ -270,7 +270,12 @@ def forward(self, observations, rnn_hidden_states, prev_actions, masks):
 
         if self.vln_config.PROGRESS_MONITOR.use and AuxLosses.is_active():
             progress_hat = torch.tanh(self.progress_monitor(x))
-            progress_loss = F.mse_loss(progress_hat.squeeze(1), observations['progress'])
+            progress_loss = F.mse_loss(
+                progress_hat.squeeze(1),
+                observations['progress'],
+                reduction='none',
+            )
+            progress_loss = (progress_loss.unsqueeze(1) * masks).sum() / masks.sum()
```
Comment (erikwijmans) [inline habitat_baselines/models/rcm/vln_rcm_policy.py]: Unfortunately these masks are not are about padding, they are episodes starting/ending (masks[t, n] == 0 denotes an episode beginning at that step).  We could use them for padding also, with the caveat that masks[0, :] will be zero instead of 1, but we would need to wire them up to do that (right now masks[1:] == 1.0 regardless of padding).

Probably the correct thing to do is give AuxLosses a (t, n) (or (t*n,)) tensor here and then extend the `reduce()` method to also take a mask.

Comment (arjunmajum) [reply habitat_baselines/models/rcm/vln_rcm_policy.py]: I fixed this for now using `weights > 0` in the `_update_agent` call.

I also used losses and masks with shape (t*n). I think I like (t, n) better so I might go back and change that at some point.
